File: ailist-core/src/main/java/me/kosik/interwalled/ailist/core/AIList.java
package me.kosik.interwalled.ailist.core;

import me.kosik.interwalled.ailist.core.model.Interval;

import java.util.ArrayList;
import java.util.Arrays;

public class AIList {

    final ArrayList<Interval> intervals;

    final long[] maxE;

    public AIList(ArrayList<Interval> intervals) {
        this.intervals = intervals;
        this.maxE = new long[intervals.size()];

        if(! intervals.isEmpty()) {
            long lastMaxE = intervals.get(0).to();

            for(int i = 0; i < intervals.size(); ++ i) {
                lastMaxE = Math.max(lastMaxE, intervals.get(i).to());
                this.maxE[i] = lastMaxE;
            }
        }
    }

    public int size() {
        return intervals.size();
    }

    public AIListIterator overlapping(final Interval interval) {
        return new AIListIterator(interval.from(), interval.to(), this);
    }

    @Override
    public String toString() {
        return "AIList{" +
                "intervals=" + intervals +
                ", maxE=" + Arrays.toString(maxE) +
                '}';
    }
}

File: ailist-core/src/main/scala/me/kosik/interwalled/ailist/core/AIList.scala
package me.kosik.interwalled.ailist.core

import me.kosik.interwalled.ailist.core.model.Interval
import java.util.ArrayList
import java.util.List


case class AIList(intervals: Array[Interval], maxE: Array[Long]) {
  
    def size(): Integer = {
        intervals.length
    }

    def overlapping(query: Interval): Array[Interval] = {
        intervals.iterator
            .zip(maxE.iterator)
            .dropWhile { case (interval, maxE) => maxE < query.from }
            .takeWhile { case (interval, maxE) => interval.from <= query.to }
            .map { case (interval, maxE) => interval }
            .filter(interval => Interval.overlaps(interval, query))
            .toArray
    }

    def overlappingJ(query: Interval): List[Interval] = {
        import scala.collection.JavaConverters._
        overlapping(query).toBuffer.asJava
    }
}

object AIList {
    // Java adapter
    def apply(intervals: ArrayList[Interval]): AIList = {
        import scala.collection.JavaConverters._
        AIList.apply(intervals.asScala.toArray)
    }

    def apply(intervals: Array[Interval]): AIList = {
        val maxE = {
            if(! intervals.isEmpty)
                intervals.scanLeft(Long.MinValue)(_ max _.to).drop(1)
            else
                Array.empty[Long]
        }

        AIList(intervals, maxE)
    }
}